include/bluetooth/bluetooth.h

```diff
@@ -398,6 +463,118 @@ int bt_le_adv_update_data(const struct bt_data *ad, size_t ad_len,
  */
 int bt_le_adv_stop(void);
 
+/** @brief Increment an advertising set`s reference count.
+ *
+ *  Increment the reference count of an advertising set object.
+ *
+ *  @param adv Advertising set object.
+ *
+ *  @return Advertising set object with incremented reference count.
+ */
+struct bt_adv *bt_adv_ref(struct bt_adv *adv);
+
+/** @brief Decrement an advertising set's reference count.
+ *
+ *  Decrement the reference count of an advertising setobject.
+ *
+ *  @param adv Advertising set object.
+ */
+void bt_adv_unref(struct bt_adv *adv);
+
+/** @brief Create advertising set.
+ *
+ *  Create a new advertising set and set advertising parameters.
+ *
+ *  @param param Advertising parameters.
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ *  @return -ECONNREFUSED When connectable advertising is requested and there
+ *			  is already maximum number of connections established.
+ *			  This error code is only guaranteed when using Zephyr
+ *			  controller, for other controllers code returned in
+ *			  this case may be -EIO.
+ */
+struct bt_adv *bt_le_adv_set_create(const struct bt_le_adv_param *param);
+
+/** Start advertising with the given advertising set
+ *
+ *  Start advertising with a specific advertising set.
+ *  If both timeout and advertising event limit is given then the advertising
+ *  set will be stopped once one of them is reached.
+ *
+ *  @param adv         Advertising set object.
+ *  @param timeout     Timeout in ms, K_FOREVER or K_NO_WAIT will disable timeout.
```

Comment:
on HCI level timeout is in 10ms units, is this going to be rounded by host eg when user pass 25ms?

Reply:
That was intention yes, not sure if it is a good idea or not. (see other comment on timeout for scanner).

Reply:
IIRC our other APIs that map to HCI exposes the HCI units to the user, but then provides convenience macros to convert from ms or seconds. I think we should try to be consistent here.

Reply:
To clarify my use of s32_t timeout here:

I added bt_scan_set_timeout(s32_t timeout), which could possibly only support advertising extension with the duration parameter, or add timeout support for legacy as well. but using k_delayed_work or something to get the same functionality.
Also since the parameter was named "timeout" I thought it best to use the timeout type s32_t.

If we only support advertising extension, then using unit is better, also I think we should avoid using the term timeout in that case, so switch back to "duration"

Reply:
Agree, I had already switched back to u16_t for scan timeout, but forgot to update the advertiser.